### DISCUSSION TITLE
horizon landscape with peakfinder or peakvisor ?

### DISCUSSION THREAD

**Discussion by dromeuf** [Stellarium/stellarium#3574] [General]
Hi, Do you know if a plugin to create a horizon with a panorama mountains identification site like peakfinder.com or peakvisor.com exists ? All the best, David.

**Comment by gzotti** [comment]
Read section 7.1.3 of the User Guide carefully.

**Reply by chrsoo** [reply]
@gzotti - so... this answer should be taken to mean "no, currently no such plugin exists, but feel free to create one?" ;-)

If one were to take on such an endeavor, is this possible using a scripting or would it have to be a plugin? Or perhaps best to start with a standalone app that creates the required landscape file as a zip?

**Reply by gzotti** [reply]
The User Guide (ah yes, actually 7.1.2) points you to HeyWhatsThat (and HoriZONE) which provides this feature. Using other data requires own work, yes. Stellarium has no GIS capabilities and does not contain any database of terrestrial mountain peaks, so it's best to use software which already has such features and data. If you create a workflow involving peakfinder/peakvisor, we also gladly accept workflow documentation for the User Guide.

**Reply by chrsoo** [reply]
Thanks - I actually _only_ read section 7.1.3 (as originally suggested) and missed the [HeyWhatsThat](https://www.heywhatsthat.com/) reference. For the next iteration of [the manual](https://stellarium.org/files/guide.pdf), it might be worth mentioning that HeyWhatsThat now supports downloading Stellarium landscape files directly.

Regarding PeakFinder, I looked a bit closer at their [Web API](https://github.com/Fabiz/PeakFinder-API), and given that it uses a Web Assembly Plugin, I have a strong suspicion that it would be (very) difficult to integrate support for PeakFinder landscapes directly in Stellarium using scripts and plugins. The solution would likely require a separate web page and changes to the PeakFinder API, but sill nothing that can realistically be done without the support of the PeakFinder author.

This said, it would be fantastic to have a PeakFinder plugin, given their smooth rendering of mountain peak outlines and if the output could somehow be rendered directly in Stellarium... the experience would be "stellar".  One can always dream... 😃

**Reply by Fabiz** [reply]
> For the next iteration of [the manual](https://stellarium.org/files/guide.pdf), it might be worth mentioning that HeyWhatsThat now supports downloading Stellarium landscape files directly.

It would be also nice to mention the PeakFinder support for creating the Horizon files. Who is in charge of writing the manual?

**Reply by gzotti** [reply]
Sure. HWT added that after horiZONE had appeared, and having just leaned about PeakFinders new abilities, I have started adding this, but now am eagerly waiting for whatever better-quality options @Fabiz can provide. I hope for "horizon polygon plus (optional?) sub-horizon rendering of foreground peaks (raster image) plus gazetteer plus metadata in description". 

The User Guide used to be a separate project years ago. However, it must be written by those who develop the code, so that it is clear what it can do. The current editors are given on the cover and have not changed in many years, more history in chapter 1.

**Reply by stephanodic** [reply]
Do you Know if i can import custom landscape in stellarium app for android? Thanks to everyone

**Comment by Fabiz** [comment]
Hi 

I’m the developer of PeakFinder and i’m happy to provide support for the horizons. 

I’ve already created some functionality to create a horizon for stellarium a few months ago:
Go to www.peakfinder.com, select a viewpoint and the choose the menu ‘Export’. There you’ll find an option to export a zip file that you may drop directly to stellarium. 

For a Api extension I would need to know a bit more about the usage numbers of Stellarium. The horizon calculations are cpu intensive. So I need to know what server load that would create.

**Reply by chrsoo** [reply]
Thank you so much - quality of the mountain skyline looks gorgeous!

Can't help you with the usage numbers, perhaps @gzotti knows more?

**Reply by gzotti** [reply]
@Fabiz this looks indeed great! However, what makes your site so interesting are the peak names. Maybe you can extend your export by: gazetteer.LANG.utf8 files, location name and location timezone (add these entries before exporting). I also like your "intermediate horizons". These would have to be exported as raster images, though. (Optional feature?)
I doubt we must create an interactive link. Clicking on our very coarse map is not enough to create a viewpoint. We either would need to create a zoomable map interface like yours or OpenStreetMap where you could select your actual patch of grass or road corner where you are observing, or the user just visits your site and creates a pano or two to download&install. It is possible to load the new landscapes into the running program, so this is perfectly fine IMHO.  I will add a note to the User Guide. 

An extension could be a dedicated Peakfinder plugin which includes the map window, requests a PF polygon, gazetteer and location data, and updates a dedicated polygonal "PeakFinder" landscape. I cannot say how many users would like to run that on a daily basis. We have about 300.000-350.000 quarterly downloads, quite a few of which like to create their own horizons, but probably only once.

**Reply by alex-w** [reply]
Probably description.en.utf8 and gazetteer.en.utf8 will be enough, because I see same English text in all description files

**Reply by Fabiz** [reply]
> Probably description.en.utf8 and gazetteer.en.utf8 will be enough, because I see same English text in all description 

This shouldn’t be a big task. I’ll see if I can add this information to the PeakFinder export in the next days.

**Comment by dromeuf** [comment]
At the time of my question, peakfinder export hadn't worked for me with version 23.4. I then used the https://www.udeuschle.de/panoramas/makepanoramas.htm site and when I searched in Chrome's F12 debug mode, I found the graphics in the /panoramas/pans directory, which I downloaded and assembled with transparent color. Here's what it looks like with a mid-season sunrise on the summit of a high mountain. 

![image](https://github.com/Stellarium/stellarium/assets/18034945/1136dda9-0c5f-48c7-887d-75924145c923)

**Reply by gzotti** [reply]
Now that looks nice. Would be even nicer if the landscape tags could be converted into the gazetteer format. And of course, if such a topography website already gives things like farthest visible distance, this meta information (and maybe some background info about the actual DTM or other data used for computation, smoothing, shading, ...) could go into the exported description, together with a link to the generator web service, when their developers can be convinced of exporting their data for use with Stellarium. Applicability for just such studies was a key motivation for fixing/extending landscapes and creating the ArchaeoLines btw.

**Reply by Fabiz** [reply]
@dromeuf just out of curiosity: what did not work with PeakFinder? Have you tried the PeakFinder-Export function for Stellarium? 
Is there something I should fix on the PeakFinder side?

**Reply by 10110111** [reply]
> what did not work with PeakFinder?

From what I see after export and what the above screenshot shows, there are at least two differences:

1. No details under the horizon (shading, contours — only the horizon itself is saved),
2. No labels of the peaks (and they don't appear when enabled in landscape settings).

**Reply by Fabiz** [reply]
Thanks for the clarifications. 

> 1. No details under the horizon (shading, contours — only the horizon itself is saved),

Adding shades and contours would be much more complex than just providing the horizon txt file with azimut/altitude. I’ve to see what I can do. 

> 2. No labels of the peaks (and they don't appear when enabled in landscape settings).

Adding this is not a lot of work. I’ll do this in the next days.

**Reply by gzotti** [reply]
re 1. The solution of @dromeuf looks great, he has apparently grabbed some intermediate data (raster images) from the browser cache. You could have an optional raster export: if activated, create an "old_style" landscape with 4 or 8 raster segments (PNG) in cylinder projection along the horizon for high resolution (try to create e.g. 2048x512 or 2048x1024 tiles for use also on limited hardware like Raspberry Pi), and combine that with the current horizon polygon data, which is still important for e.g. sunrise computations. The bottom part can be closed by a tiny groundtex (just use a 2x2px light gray to close), or define an image which could be a decorative wind rose or something more interesting, e.g. a viewshed graph that indicates viewing distance).

If you base your computations on some particular atmospheric pressure/temperature values, you should write them to the landscape.ini file to recreate a fitting amount of refraction. 

Please allow users to name the respective landscape, use that name also in the downloadable file, and maybe add an optional description box, so that some info about what makes this viewpoint so special could be given by the landscape creator before the ZIP file is created and possibly distributed without such info. Else, it's up to you to write something more interesting than a creation date in the description, e.g. add a link to your website of course, or notify users of any usage restrictions on the data if required.

**Reply by dromeuf** [reply]
@Fabiz your site is great but the export is the same description as @10110111 . Appearance gave nothing away in stellarium. If you can complete it a bit like my example, that would be perfect.

**Reply by Fabiz** [reply]
@gzotti Your idea with the separate png segments could work quite easily. On peakfinder.com there is also an option for a highres svg output. In this svg file only the peak lines and labels are vector graphics. The background consists of separate bitmap segments. I think I can use these bitmaps also for Stellarium. I’ll check this next week.

**Comment by dromeuf** [comment]
example parameters for this pano with UDEUSCHLE 
![Capture config horizon Est](https://github.com/Stellarium/stellarium/assets/18034945/3c2409ea-3eed-4583-9c1f-69e617207160)

**Reply by dromeuf** [reply]
and result with GIMP mount PNG 21600 x 750 pixels

![HorizonSanctuaireCorentUDEUSCHLE-AvecNomsSommets-21600x750](https://github.com/Stellarium/stellarium/assets/18034945/3509ff3b-b98d-4131-bc56-c750d024e3b7)

**Reply by gzotti** [reply]
Yes, such large panos must be broken into e.g. 12 pieces (to be below 2048) please to be usable on older systems.

**Comment by Fabiz** [comment]
I've just released an update for Stellarium landscape exports on [www.peakfinder.com](https://www.peakfinder.com).
On the export dialog the following options are available:

- Output format Bitmap (Textures) or Vector (Horizonline)
- Style: Dark or Light
- Resolution of the textures: 512px, 1024px, 2048px
- Custom name

![PeakFinder Horizon for Stellarium](https://github.com/user-attachments/assets/140ef030-505f-46f5-86cc-a6fc1423153a)

**Reply by 10110111** [reply]
Great work! I've just tested the real Stellarium landscape generated by PeakFinder, and it looks mostly great—until I zoom in, which reveals that the lines are aliased. If you could export an antialiased (say, supersampled) version, this would result in a much better quality at the same image resolution.

**Reply by Fabiz** [reply]
I see what you mean. thanks for the feedback. I'll try to improve this in the next days.

**Reply by Fabiz** [reply]
I've added supersampling to the newest version. The output looks better now. 
Thanks for the hint and your support.

**Comment by Fabiz** [comment]
Here an example of a night view from Oberrothorn:

![PeakFinder view from Oberrothorn night](https://github.com/user-attachments/assets/b0fcf2bb-bedb-44f2-9312-dab1fd7a6739)

**Comment by Fabiz** [comment]
and here a day view: 

![PeakFinder view from Oberrothorn day](https://github.com/user-attachments/assets/1869f1e1-f62b-45d3-953d-3b5d387d5219)

**Comment by 10110111** [comment]
Not sure where to report this, but the following location has a strange
view to the east: as if the terrain has holes. Location: 20.40773°S,
57.36892°E (Case Noyale, Mauritius).

**Reply by Fabiz** [reply]
I'm not sure what you mean. For me the view looks ok.
https://www.peakfinder.com/?lat=-20.40773&lng=57.36892&azi=80.78&alt=6.08&fov=110&cfg=s&name=20.40773%c2%b0S%2c%2057.36892%c2%b0E

Can you send a picture of the error directly to info@peakfinder.com ?

**Reply by 10110111** [reply]
Hmm, must have been a glitch of Chrome on Android. Now I no longer reproduce it anywhere.

**Comment by gzotti** [comment]
@Fabiz Comparing this with other workflows the landscape seemed to be too low. I then activated the polygon line and see this does not align with the top line of raster image. Some sections are higher, some are lower. Are you sure about the geometry of the rasters: Are these perspective views or do they represent azimuth/altitude along 360 degrees? Are the angles OK?
... Ah no, two errors, and a third change follows. You need:


```
calibrated = true
decor_angle_shift = -22.5
ground_angle_shift = -22.5

```

And in [location] please add another 
`name = <placename>
`
so it will be used&displayed in the status line. 

Still, is the observer placed on the ground (say, 1.5-2m above DSM), or something like 5m above ground? The near horizon seems always too favourable.

**Reply by Fabiz** [reply]
@gzotti I'm having problems to understand how Stellarium translates the textures passed by old_style to the horizon.

I've created the following test setup to visualize this issue:
The landscape file [PeakFinder.zip](https://github.com/user-attachments/files/17488677/PeakFinder.zip) contains 8 512x512 textures each with a horizontal red line in the middle of the image. 
The landscape.ini contains the following information:

```
nbsidetex = 8
nbside = 8
tex0 = peakfinder0.png
side0 = tex0:0:0:1:1
...
nb_decor_repeat = 1
decor_alt_angle = 45
decor_angle_shift = -22.5
```

Additionally I've added a black line to horizon.txt at an altitude of 0°.
With this setup I'd expect that the red line would match exactly the 0° altitude (black line). But this is not the case. The red line is displayed in Stellarium at -1°41'. 

![Texture Offset](https://github.com/user-attachments/assets/be4c73aa-d56e-4ff9-94c2-3ba966b856a6)


Have I overseen something?

**Reply by gzotti** [reply]
As written above. My first fix in Stellarium, around 2009, was finding an error in the vertical alignment. Given existing hand-massaged horizons, I was required to add a flag, `calibrated`, which you mandatorily must set if you use the numbers.

**Reply by Fabiz** [reply]
@gzotti Thanks for the hint. Adding the calibrate flag fixed these issues (you've mentioned the flag before but I've missed it, sorry for this).
The newest version on peakfinder.com seems to be fine now. The horizon.txt is also referenced.

**Reply by gzotti** [reply]
Excellent. Angled gazetteer labels are available in a pull request (i.e. coming soon). (You could also add back the peak altitudes!) Unfortunately, the sub-zero angle problem is not solved yet (seems it turned up last year, but I have not git bisected yet). So, please, for now, use the raster image for "real"  image and max(0, alt) for altitudes in the polygon definition. (Users can also then switch off the polygon line.)

**Reply by dromeuf** [reply]
Hi,

Please, can you explain when calibrated=true is absolutely necessary?

In this case for example ?

```
[landscape]
name = Puy-de-Dôme Horizon with names https://www.udeuschle.de/panoramas
author = dromeuf
description = Horizon depuis le sommet du Puy-de-Dôme avec les noms des sommets
type = spherical
maptex = HorizonSommetPuyDeDome-UDEUSCHLE-WithNames-21600x750.png
maptex_top=6.25
maptex_bottom=-6.25
angle_rotatez=-90

[location]
light_pollution = 1
atmospheric_extinction_coefficient = 0.2
atmospheric_temperature = 10
atmospheric_pressure = -1
planet = Earth
latitude = +45d46'20.8"
longitude = +2d57'52.2"
altitude = 1465
timezone = Europe/Paris

```

**Reply by gzotti** [reply]
As described in the User Guide, it is absolutely necessary when you want correct values in old_style landscapes, unless you are using a pre-2009 hand-adjusted decoration. Note that the default Guereins landscape is not calibrated.

**Reply by dromeuf** [reply]
ok so, as the documentation in 7.1.4 specifies, an old style is a landscape cut into several pieces using decor_alt_angle. You use calibrated=true to adjust the h altitude. So in my case, where I've used udeuschle above with only 1 global image with the maptex_top and bottom specified in the landscape.ini, I don't need to specify calibrate=true. Kind regards.

**Comment by ArchaeoastronomyIreland** [comment]
Great work Fabiz !

A couple of things

I think there still may be a slight error with the placement of the horizon rasters (I will see if I can quantify this when I have more time)
Horizon.txt not showing negative horizon values.

Kills stellarium on systems with old graphics cards (......time for a new laptop for me)

Suggestion for Georg (At the landscape load make the gazetteer optional)

**Reply by gzotti** [reply]
The horizon.txt having issues with negative altitudes is a newly reported issue. It used to work. Not sure yet when this crept in. 
Old GPU: how old is old? the textures are 1024 or 2048px tiles, this should be compatible with everything still powering up and supported by current OS/drivers.
Gazetteer: Just switch off the landscape labels in the landscape GUI.

**Comment by Fabiz** [comment]
> Kills stellarium on systems with old graphics cards (......time for a new laptop for me)

can you try to uncomment the horizon.txt file and just use the textures? I had also some framerate drops even on with a modern gpu.

**Reply by ArchaeoastronomyIreland** [reply]
Hi Fabio, apologies for the late response. Yes I will try this and update.
